packages/typescript-react-native-resolver/src/resolve.ts

```diff
@@ -0,0 +1,205 @@
+import {
+  FileModuleRef,
+  getMangledPackageName,
+  PackageModuleRef,
+} from "@rnx-kit/tools-node";
+import path from "path";
+import ts from "typescript";
+
+import { findModuleFile } from "./module";
+import { ResolverContext } from "./types";
+
+/**
+ * Resolve a module reference within a given package directory.
+ *
+ * If a module path is given, use that to find the corresponding module
+ * file.
+ *
+ * Otherwise, consult `package.json` for properties which refer to
+ * "entry points" within the package (e.g. `types`, `typings` and `main`).
+ * If those properties don't resolve the module, then fall back to looking
+ * for an "index" file.
+ *
+ * @param context Resolver context
+ * @param packageDir Root of the package which contains the module
+ * @param modulePath Optional relative path to the module
+ * @param extensions List of allowed file extensions, in order from highest precedence (index 0) to lowest.
+ * @returns Resolved module, or `undefined` if resolution fails
+ */
+export function resolveModule(
+  context: ResolverContext,
+  packageDir: string,
+  modulePath: string | undefined,
+  extensions: ts.Extension[]
+): ts.ResolvedModuleFull | undefined {
+  //  A module path was given. Use that to resolve the module to a file.
+  if (modulePath) {
+    return findModuleFile(context, packageDir, modulePath, extensions);
+  }
+
+  const { host, log } = context;
+
+  let module: ts.ResolvedModuleFull | undefined;
+
+  //  No path was given. Try resolving the module using package.json
+  //  properties.
+  log.log("Reading package.json from directory %o.", packageDir);
+  const result = host.readFile(path.join(packageDir, "package.json"));
+  if (!result) {
+    throw new Error(
+      `Failed to read package.json from directory '${packageDir}'`
+    );
+  }
+  const { types, typings, main } = JSON.parse(result);
+
+  //  Only consult 'types' and 'typings' properties when looking for
+  //  type files (.d.ts).
+  if (extensions.includes(ts.Extension.Dts)) {
+    if (typeof types === "string" && types.length > 0) {
+      log.log("Package has 'types' field '%s'.", types);
+      module = findModuleFile(context, packageDir, types, extensions);
+    } else if (typeof typings === "string" && typings.length > 0) {
+      log.log("Package has 'typings' field '%s'.", typings);
+      module = findModuleFile(context, packageDir, typings, extensions);
+    }
+  }
+  if (!module && typeof main === "string" && main.length > 0) {
+    log.log("Package has 'main' field '%s'.", main);
+    module = findModuleFile(context, packageDir, main, extensions);
+  }
```

Comment:
If we need to look at more fields, I think something like this would be better:
```suggestion
  const manifest = JSON.parse(result);

  [
    // Only consult 'types' and 'typings' properties when looking for
    // type files (.d.ts).
    ...(extensions.includes(ts.Extension.Dts) ? ["types", "typings"] : []),
    "main",
  ].find((field) => {
    const modulePath = manifest[field];
    if (typeof modulePath === "string" && modulePath.length > 0) {
      log.log("Package has '%s' field '%s'.", field, modulePath);
      module = findModuleFile(context, packageDir, modulePath, extensions);
      return Boolean(module);
    }
    return false;
  });
```

Reply:
Makes sense. Adding more fields would benefit from rolling up into a loop.